src/libclipper/src/selection_policies.cpp

```diff
@@ -1,344 +1,484 @@
+  return state.debug_string();
+};
 
-SimpleState SimplePolicy::initialize(
-    const std::vector<VersionedModelId>& candidate_models) {
-  // TODO: IMPLEMENT
-  assert(candidate_models.size() > 0);
-  return SimpleState(candidate_models);
-}
+//// ********
+//// * EXP4 *
+//// ********
 
-SimpleState SimplePolicy::add_models(SimpleState state,
-                                     std::vector<VersionedModelId> new_models) {
-  state.insert(state.end(), new_models.begin(), new_models.end());
-  return state;
+PolicyState Exp4Policy::initialize(
+    const std::vector<VersionedModelId>& candidate_models_) {
+  return Exp3Policy::initialize(candidate_models_);
 }
 
-long SimplePolicy::hash_models(
-    const std::vector<VersionedModelId>& candidate_models) {
-  UNUSED(candidate_models);
-  return 0;
+PolicyState Exp4Policy::add_models(
+    PolicyState state, const std::vector<VersionedModelId>& new_models) {
+  return Exp3Policy::add_models(state, new_models);
 }
 
-std::vector<PredictTask> SimplePolicy::select_predict_tasks(SimpleState state,
-                                                            Query query,
-                                                            long query_id) {
-  std::vector<PredictTask> task_vec;
-
-  // construct the task and put in the vector
-  for (auto v : state) {
-    task_vec.emplace_back(query.input_, v, 1.0 / (float)state.size(), query_id,
-                          query.latency_micros_);
+std::vector<PredictTask> Exp4Policy::select_predict_tasks(PolicyState& /*state*/,
+                                                Query query,
+                                                long query_id) {
+  // Pass along all models selected
+  std::vector<PredictTask> tasks;
+  for (VersionedModelId id : query.candidate_models_) {
+    auto task =
+        PredictTask(query.input_, id, 1.0, query_id, query.latency_micros_);
+    tasks.push_back(task);
   }
-  return task_vec;
+  return tasks;
 }
 
-Output SimplePolicy::combine_predictions(SimpleState state, Query query,
-                                         std::vector<Output> predictions) {
-  UNUSED(state);
-  UNUSED(query);
-  // just return the first prediction
-  if (predictions.empty()) {
-    return Output{0.0, std::make_pair("none", 0)};
-  } else {
-    float sum = 0;
-    for (auto o : predictions) {
-      sum += o.y_hat_;
-    }
-    return Output{sum, std::make_pair("all", 0)};
+Output Exp4Policy::combine_predictions(PolicyState state, Query /*query*/,
+                                  std::vector<Output> predictions) {
+  // Weighted Combination of All predictions
+  auto y_hat = 0;
+  std::vector<VersionedModelId> models;
+  for (auto p : predictions) {
+    auto model_id = (p.models_used_).front();
+    y_hat += (state.model_map_[model_id]["weight"] / state.weight_sum_) * p.y_hat_;
+    models.push_back(model_id);
   }
+  auto output = Output(y_hat, models);
+  return output;
 }
```

Comment:
I'm not sure what you're asking? As the code is right now, it won't work for binary classification. It will work for regression where the true average is a useful prediction. We want it to work for at least binary classification.